File: MSc in Data Science/Semester 1/Data Structures and Algorithms/Menu Driven Array Operations/ARRAY.cpp
/////////////Program for Array Operations ///////////////
#include<stdio.h>
#include<stdlib.h>
int a[1000], pos, elem ;
int n = 0;
void create();
void display();
void insert();
void del();
void sort();
void search();

int main()
{

            int choice;

            while(1)

            {

             printf("\n\n\n >>>MENU<<<\n\n\n");

             printf("\n Press 1 to create an array. \n");

             printf("\n Press 2 to traverse/print the array. \n");

             printf("\n Press 3 to insert an element at a desired position. \n");

             printf("\n Press 4 to delete an element at a given position. \n");
             
			 printf("\n Press 5 to search an element inside the array. \n");
             
			 printf("\n Press 6 to sort the array in ascending order. \n");

             printf("\n Press 0 to exit \n");

             printf("\n\n\n Please enter your choice: ");

             scanf("%d", &choice);
           
///////////////////////////////////////////////////////////////
          
		   switch(choice)

           {

            case 1: create();

                  break;

            case 2: display();

                  break;

            case 3: insert();

                  break;

            case 4: del();

                  break;
                  
            case 5: search();
           
					break;
					
			case 6: sort();
			
					break;
							
            case 0: exit(0);

                  break;
                  
        default:printf("\n Please enter a valid choice:");

                        }

            }

}
/////////////////////////////////////////////////////////////

 void create()

{

            int i;

            printf("\n Enter the length of the array: ");

            scanf("%d", &n);

            printf("\n Enter %d elements in the array sepeprated by 'space' or 'enter':\n ",n);

            for(i=0; i<n; i++)

            {

                        scanf("%d", &a[i]);

            }

}

void display()

{

            int i;

            if(n == 0)

            {

                        printf("\n No elements to display");

                        return;

            }

            printf("\n Array elements are: ");

            for(i=0; i<n;i++)

                        printf("%d\t ", a[i]);

}
///////////////////////////////////////////////

void insert()

{

            int i;

            if(n == 1000)

            {

        printf("\n Array is full. Insertion is not possible");

                        return;

            }

            do

            {

             printf("\n Please enter a valid position where element to be inserted:    ");

             scanf("%d", &pos);

            }
			
			while(pos > n);

            printf("\n Enter the value to be inserted:   ");

            scanf("%d", &elem);

            for(i=n-1; i>=pos ; i--)

            {

                        a[i+1] = a[i];

            }

            a[pos] = elem;

            n = n+1;

            display();

}

///////////////////////////////////////////

void del()

{

            int i;

           

            if(n == 0)

            {

                        printf("\nArray is empty and no elements to delete");

                        return;

            }

            do

            {

                        printf("\n Please enter a valid position from where element to be deleted:    ");

                        scanf("%d", &pos);

            }while(pos>=n);

            elem = a[pos];

            printf("\n The deleted element is : %d \n", elem);

            for( i = pos; i< n-1; i++)

            {

                        a[i] = a[i+1];

            }

            n = n-1;

            display();

}
//////////////////////////////////////////////////////////////
 
void search()

{
	int i,srch;
	
	printf("Please enter a number you want to search in the array \n");
    
	scanf("%d", &srch);

     for (i = 0; i < n; i++)
      {
        if (a[i] == srch)    
          {
            printf("%d is present at location %d\n", srch, i+1);
            
            break;
          }
      }
      
    if (i == n)
  
      printf(" \n Sorry %d isn't present in the array!\n", srch);

}
////////////////////////////////////////////////////////

void sort()
{
	int i,j,temp;
	

	
	  for (i = 0; i < n; ++i) 
        {
 
            for (j = i + 1; j < n; ++j)
            {
 
                if (a[i] > a[j]) 
                {
 
                    temp =  a[i];
                    
                    a[i] = a[j];
                    
                    a[j] = temp;
 
                }
 
            }
 
        }
 
        printf("The elements of the array afterbeing sorted in ascending order are : \n");
        
		for (i = 0; i < n; ++i)
           
		    printf("%d\t", a[i]);
 
    }

File: MSc in Data Science/Semester 1/Data Structures and Algorithms/Menu Driven Array Operations/array.py
# Array
# Menu Driven Function

#insert function
def insert():
    global x
    global a
    x = int(input("Enter the number of elements you wish to insert: "))
    i = a
    if (x+a) > n :
        print("Not Possible.")
    else:
        while i < (a+x) :
            y = int(input("Enter the element one by one: "))
            global MENU
            MENU[i] = y
            i += 1
        b = x+a
        a = b

#search function
def search():
    y = int(input("What do you wish to search for?  "))
    found = False
    i = 0
    while i < a:
        if MENU[i] == y:
            found = True
            print(f"{y} has been found at {i + 1}th position")
        i += 1
    if not found :
        print("Not Found.... \n")

#delete function
def deletion():
    global MENU
    global a
    y = int(input("Which element do you wish to delete?  "))
    i = 0
    while i < a:
        if MENU[i] == y:
            j = i
            while j < a-1:
                MENU[j] = MENU[j+1]
                j += 1
            MENU[j] = 0
            a = j
        i += 1

#sorting function (ascending)
def sort():
    global MENU
    for i in range(1,a):
        temp = MENU[i]
        j = i - 1
        while j >= 0 and temp < MENU[j] :
            MENU[j+1] = MENU[j]
            j -= 1
        MENU[j+1] = temp


print("MENU Driven Program")
n = int(input("Enter the number of spaces to be declared:  "))
MENU = [0]*n
a = 0
should_continue = True
while should_continue:
    order = int(input("Type 1 for Insertion\n2 for Deletion\n3 for Search\n4 for Sorting\n5 for Traversing\n0 to Exit\n"))
    if order == 1:
        insert()
    elif order == 2:
        deletion()
    elif order == 3:
        search()
    elif order == 4:
        sort()
    elif order == 5:
        print(MENU)
    elif order == 0:
        should_continue = False
    else:
        print("Wrong choice! Please Try Again.")


